Add support for learning new commands

felipediel opened the pull request:
This update creates a generic service in the 'remote' component to enable remote control platforms to learn new commands.


## Description:


**Related issue:** [#236](https://github.com/home-assistant/architecture/issues/236)

### Example 1: Learn a single command
```yaml
- service: remote.learn_command
  data:
    entity_id: remote.bedroom
    device: television
    command: 'Turn on'
```
### Example 2: Learn a list of commands
```yaml
- service: remote.learn_command
  data:
    entity_id: remote.bedroom
    device: television
    command:
      - 'Turn on'
      - 'Turn off'
      - 'Source'
      - 'Channel up'
      - 'Channel down'
      - 'Volume up'
      - 'Volume down'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html

felipediel committed:
Add support for learning new commands

This update creates a generic service in the 'remote' component to enable remote control platforms to learn new commands.

pvizeli reviewed (commented):
For which platform will that be later?

EDIT: I saw it on architecture issue :)

pvizeli commented:
Please add a `supported_feature` functionality for that. I.e. Light/Climate/MediaPlayer

felipediel commented:
> Please add a `supported_feature` functionality for that. I.e. Light/Climate/MediaPlayer

Hello friend. Do you mean something like this?
```yaml
REMOTE_SERVICE_LEARN_COMMAND_SCHEMA = REMOTE_SERVICE_SCHEMA.extend({
    vol.Optional(ATTR_DEVICE): cv.string,
    vol.Optional(ATTR_COMMAND): vol.All(cv.ensure_list, [cv.string]),
    vol.Optional(ATTR_ALTERNATIVE): cv.boolean,
    vol.Optional(ATTR_TIMEOUT): cv.positive_int,
    vol.Optional(ATTR_SUPPORTED_FEATURES): vol.Any('Light', 'Climate', 'MediaPlayer')
})
````
Okay, I'll do it, just tell me the types that it needs to support, or if you want me to simply validate a string and leave for the platforms to choose.

Just out of curiosity, how will you implement this functionallity inside the remote control platforms?

I ask this because I'm developing a platform and I have the following JSON format in mind to store the codes:
```json
{
    "Bedroom TV": {
        "type": "Television",
        "brand": "Philips",
        "model": "42PFL3403",
        "commands": {
            "Turn on": "JgA0ABsbHRo4Gh0aHBobHB0ZHRobNx0aOBocAA0F==",
            "Turn off": "JgA0ABscNhwbGxscGxsbHBscGxsbNxscNhwbAA0F=="
        }
    }
}
```
Here I've opted to include "type", "brand" and "model" information. These labels are not directly useful for sending and receiving codes, but I believe they are important for users to share their codes in the future and keep their collection of codes organized. But I decided not to receive these header entries through the learn_command service for two reasons:

1. These information only needs to be set once per device, and I believe the best way would be through config entries or some kind of data flow. It would be a huge task to repeat the brand and model for every code the user wants to learn.

2. This approach may lead to conflicts, let's assume that the user sends this payload:
```python
- service: remote.learn_command
  data:
    - device: 'Bedroom TV'
      type: 'Television'
      brand: 'Philips'
      model: '42PFL3403'
      command: 'Turn on'
```
Followed by this another one:
```python
- service: remote.learn_command
  data:
    - device: 'Bedroom TV'
      type: 'Television'
      brand: 'Philips'
      model: '52PFL3404'
      command: 'Turn off'
```
The models conflict for the same device, which one should be stored in the JSON file?

So maybe this also aplies to 'supported_features', what if I tell it's a 'MediaPlayer' when I capture the first command and then tell this is a 'Light' when I capture the second. How should platforms handle it?

I think storing the device type is a great idea to keep the codes organized, but maybe this information should be captured elsewhere and only once per device.

Alright, I've made my point, but it's up to you to decide. Just tell me and I'll do it.

pvizeli commented:
I mean some things like that: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py#L34-L41

MartinHjelmare commented on homeassistant/components/remote/__init__.py:
```diff
@@ -110,3 +125,15 @@ def async_send_command(self, command, **kwargs):
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        return self.hass.async_add_job(ft.partial(
```
Use `await self.hass.async_add_executor_job`, instead of returning the task.

felipediel replied:
Like this?

```py
def async_learn_command(self, **kwargs):
    """Learn a command from a device.

    This method must be run in the event loop.
    """
    await self.hass.async_add_executor_job(ft.partial(
        self.learn_command, **kwargs))
```

Okay, I'll update it, but what's the idea behind it? I got this snippet from async_send_command, should I change it as well?

MartinHjelmare replied:
We can change async_send_command separately. We define coroutines for the async method interface.

felipediel replied:
Done.

MartinHjelmare commented on homeassistant/components/remote/__init__.py:
```diff
@@ -110,3 +125,15 @@ def async_send_command(self, command, **kwargs):
+    def async_learn_command(self, **kwargs):
+        """Learn a command from a device.
+
+        This method must be run in the event loop and returns a coroutine.
```
Update the docstring. We shouldn't return a coroutine. We can remove the last line.

felipediel replied:
Done.

felipediel commented:
> I mean some things like that: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py#L34-L41

Got it! What do you think of adding the following lines to the RemoteDevice class in the `__init__.py` file:

```py
@property
def supported_features(self):
    """Return the list of supported features."""
    raise NotImplementedError()
```

I think it would be interesting to define some constants to be imported by platforms that will use this property, but I do not know what kind of features you want to support? Please give me a list of some features you would like to support at first and I will add the constants to the code.

MartinHjelmare commented on homeassistant/components/remote/__init__.py:
```diff
@@ -110,3 +125,15 @@ def async_send_command(self, command, **kwargs):
+        """Learn a command from a device."""
+        raise NotImplementedError()
+
+    def async_learn_command(self, **kwargs):
```
This should be made a coroutine, using `async def`.

pvizeli replied:
no, because we return a coroutine/task object

MartinHjelmare replied:
No we shouldn't use that pattern anymore. We should define coroutines and await the task inside them.

felipediel replied:
Done.

pvizeli commented:
@felipediel 

Default is: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py#L480-L483

Your new feature is that what we need as `SUPPORT_LEARN_COMMAND`

felipediel committed:
Update __init__.py with the proposed changes

- Add 'supported_features' property and a constant related to the 'learn_command' functionality.
- Redefine 'async_learn_command' function as a coroutine.

felipediel commented:
> @felipediel
> 
> Default is: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py#L480-L483
> 
> Your new feature is that what we need as `SUPPORT_LEARN_COMMAND`

Now I see the full picture. Done.

pvizeli committed:
Update __init__.py

felipediel committed:
Fix assertion error

Adding the 'supported_features' attribute generated an assertion error on the 'Demo Remote' platform. This update fixes this.

felipediel committed:
Merge branch 'learn-command' of https://github.com/felipediel/home-assistant into learn-command

felipediel committed:
Fix duplicated 'hass' object

This update fixes a typo that occurred at the last update.
